{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Копия Zarubin_HW_2_Algorithms_in_bioninformatics",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8ZRgX-b3DzY_"
      },
      "source": [
        "# Домашнее Задание 2\n",
        "НИУ ВШЭ, Зарубин Дмитрий"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5SaZSt_GD22T"
      },
      "source": [
        ">#### 1. Скачать последовательность генома E.coli."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "__Zts4SxDoZI",
        "outputId": "8ad2da2c-8b10-45d4-8afd-4a178296157f"
      },
      "source": [
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "!gunzip GCF_000005845.2_ASM584v2_genomic.fna.gz"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-03-11 10:36:27--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.11, 2607:f220:41e:250::13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1379902 (1.3M) [application/x-gzip]\n",
            "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’\n",
            "\n",
            "GCF_000005845.2_ASM 100%[===================>]   1.32M  2.51MB/s    in 0.5s    \n",
            "\n",
            "2021-03-11 10:36:28 (2.51 MB/s) - ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’ saved [1379902/1379902]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5BstLxPjtGMy",
        "outputId": "6305ccd0-96c5-4679-921d-0d855408011e"
      },
      "source": [
        "pip install biopython"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting biopython\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/3a/cd/0098eaff841850c01da928c7f509b72fd3e1f51d77b772e24de9e2312471/biopython-1.78-cp37-cp37m-manylinux1_x86_64.whl (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 10.9MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.19.5)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.78\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "weXsMyDyEZWI"
      },
      "source": [
        "import regex as re\n",
        "import itertools\n",
        "from tqdm import tqdm\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from math import log\n",
        "from statistics import mean\n",
        "from random import randrange\n",
        "from Bio import SeqIO"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0l2br7vwJgqK"
      },
      "source": [
        ">#### 2. Выделить 1000 фрагментов размером 4000 нуклеотидов. Можно с фиксированным шагом, можно с помощью датчика случайных чисел."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "A8Kwg9TCtReB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "45d54352-98c4-4081-8dc9-2e971ffdbc60"
      },
      "source": [
        "# Читаем последовательность из файла\n",
        "genome = str(SeqIO.read('GCF_000005845.2_ASM584v2_genomic.fna','fasta').seq)\n",
        "print(f'Длина генома | {len(genome)}')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Длина генома | 4641652\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "rrl-N_QoImDA"
      },
      "source": [
        "# Делим геном на 1000 последовательностей, каждая длинной 4000 н.\n",
        "# Случайно бросаем точку в интервале [0; длина генома - 4000]\n",
        "# Берем фрагмент в интервале [случайная точка; случайная точка+4000]\n",
        "\n",
        "\n",
        "sequences = [[]for _ in range(1000)]# двойной массив размера (1000, 4000), \n",
        "                                    # в котором хранятся все фрагменты.\n",
        "for i in range(1000):\n",
        "    r = randrange(len(genome)-4000)\n",
        "    sequences[i] = genome[r:r+4000]\n",
        "sequences = np.array(sequences)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bZf7jEOcKtoP"
      },
      "source": [
        ">#### 3. Для каждого фрагмента построить бернулиевскую модель и марковские модели 1,2,3 порядка"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_4yf6hV6EodE"
      },
      "source": [
        "# Функция для подсчета частот для модели k-порядка.\n",
        "\n",
        "def frequences(sequence, k=0):\n",
        "    bases = ('A', 'C', 'T', 'G')\n",
        "    freqs = {} # Пустой словарь\n",
        "    # считаем частоты для каждого k+1-мера последовательности.\n",
        "    \n",
        "    for i in range(k, k+2):# perm - пербор всех возможных комбинаций из bases с длиной от k+1 до k+2\n",
        "        \n",
        "        freqs[i] = {}\n",
        "        for perm in list(itertools.product(bases, repeat=i)):\n",
        "            freqs[i][''.join(perm)] = len(re.findall(''.join(perm), sequence, overlapped=True))/\\\n",
        "                                  (len(sequence) - k)\n",
        "                                  \n",
        "    return freqs"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D4AVHr-XGbD-",
        "outputId": "e7ff1f2c-b327-4e9e-873a-3b645928966d"
      },
      "source": [
        "frequences(sequences[0], k=1)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{1: {'A': 0.26206551637909475,\n",
              "  'C': 0.2528132033008252,\n",
              "  'G': 0.25406351587896975,\n",
              "  'T': 0.23130782695673918},\n",
              " 2: {'AA': 0.09177294323580895,\n",
              "  'AC': 0.053513378344586145,\n",
              "  'AG': 0.0622655663915979,\n",
              "  'AT': 0.05451362840710178,\n",
              "  'CA': 0.07001750437609403,\n",
              "  'CC': 0.05901475368842211,\n",
              "  'CG': 0.07126781695423856,\n",
              "  'CT': 0.05251312828207052,\n",
              "  'GA': 0.060515128782195546,\n",
              "  'GC': 0.08152038009502376,\n",
              "  'GG': 0.06001500375093773,\n",
              "  'GT': 0.052013003250812706,\n",
              "  'TA': 0.03975993998499625,\n",
              "  'TC': 0.0587646911727932,\n",
              "  'TG': 0.060515128782195546,\n",
              "  'TT': 0.07201800450112528}}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Z3eV4qxMl40-"
      },
      "source": [
        "# Функция для подсчета начальных вероятностей - q(марковских моделей)\n",
        "def starting_freqs(dataframe, k=1):\n",
        "    bases = ('A', 'C', 'T', 'G')\n",
        "    freqs = {} # Пустой словарь\n",
        "    # считаем частоты для префиксов последовательности\n",
        "    # создаем словарь для хранения частот\n",
        "    for i in range(1, k+2):# perm - пербор всех возможных комбинаций из bases с длиной от 1 до k+1\n",
        "        for perm in list(filter(lambda x: len(x) > 0, itertools.product(bases, repeat=i))):\n",
        "            freqs[''.join(perm)] = 0\n",
        "        # пробегаем по каждому фрагменту из генома\n",
        "        # берем отрезок от 0 до k и считаем для него вхождения\n",
        "        for array in dataframe:\n",
        "            freqs[array[0:i]] += 1\n",
        "    # считаем частоты\n",
        "    for item in freqs:\n",
        "        freqs[item] = freqs[item]/1000\n",
        "    return freqs"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "870wsK4VF55c",
        "outputId": "16833156-9e10-41ae-c4b6-9dc4a3550fce"
      },
      "source": [
        "starting_freqs(sequences, k=1)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'A': 0.265,\n",
              " 'AA': 0.065,\n",
              " 'AC': 0.067,\n",
              " 'AG': 0.061,\n",
              " 'AT': 0.072,\n",
              " 'C': 0.274,\n",
              " 'CA': 0.057,\n",
              " 'CC': 0.072,\n",
              " 'CG': 0.081,\n",
              " 'CT': 0.064,\n",
              " 'G': 0.236,\n",
              " 'GA': 0.057,\n",
              " 'GC': 0.085,\n",
              " 'GG': 0.046,\n",
              " 'GT': 0.048,\n",
              " 'T': 0.225,\n",
              " 'TA': 0.048,\n",
              " 'TC': 0.042,\n",
              " 'TG': 0.067,\n",
              " 'TT': 0.068}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 83
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "twtNVqPTZNID"
      },
      "source": [
        "# Рассчет BIC AIC\n",
        "\n",
        "def calc_BIC (L,s,n=1000):\n",
        "    # s - кол-во независимых параметров модели\n",
        "    return s*log(n) - 2*L\n",
        "\n",
        "\n",
        "def calc_AIC (L, s):\n",
        "    return 2*s - 2*L"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GHav6gFxZdaJ"
      },
      "source": [
        "# Функция возвращает словарь со скорами AIC и BIC для модели Бернулли.\n",
        "\n",
        "def Bernoulli (sequence, n=1000, k=0):\n",
        "    freqs = [[]for _ in range(n)]# Двойной массив размера (1000, 4^k+1) для частот\n",
        "    BIC = 0# Двойной массив размера (1000, 1) для BIC\n",
        "    AIC = 0# Двойной массив размера (1000, 1) для AIC\n",
        "    s = 4**(k + 1) - 1 # Число параметров\n",
        "    for i in range(n):\n",
        "        L = 0 # loglikelihood\n",
        "        freqs[i] = frequences(sequences[i], k=k)[k+1]# для каждого фрагмента считаются частоты (k+1)мера\n",
        "        for nucleotide in sequences[i]:\n",
        "            L += log(freqs[i][nucleotide])# бежим по последовательности и \n",
        "                                          # прибавляем логарифм часоты для каждого символа. \n",
        "        # Считаем AIC BIC для каждого фрагмента.\n",
        "        AIC += calc_AIC(L, s)\n",
        "        BIC += calc_BIC(L, s)\n",
        "    return {'Model' : 'Bernoulli',\n",
        "            'AIC' : round(AIC),\n",
        "            'BIC' : round(BIC)}"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1EFJIo0141pm"
      },
      "source": [
        "# Функция возвращает словарь со скорами AIC и BIC для Марковской модели k-порядка.\n",
        "\n",
        "def Markov(sequence, n=1000, k=1):\n",
        "    freqs = [[]for _ in range(n)]# Двойной массив размера (1000, 4^k+1) для частот\n",
        "    BIC = 0# Двойной массив размера (1000, 4000) для BIC\n",
        "    AIC = 0# Двойной массив размера (1000, 4000) для AIC\n",
        "    s = 4**(k + 1) - 1 # Число параметров модели\n",
        "    q = starting_freqs(sequence, k=k)# Считаем начальные частоты k-меров\n",
        "    # Для каждой послед-ти\n",
        "    for i in range(n):\n",
        "        L = 0\n",
        "        # прибавляем логарифмы начальных частот \n",
        "        for t in range(1,k+1):\n",
        "            L += log(q[sequence[i][0:t]])\n",
        "        # Считаем переходные вероятности\n",
        "        freqs[i] = frequences(sequences[i], k=k)\n",
        "        for base in list(freqs[i][k+1]):\n",
        "            # если наткнемся на не встерчающийся k-мер\n",
        "            if freqs[i][k][base[:-1]] == 0:\n",
        "                pass\n",
        "            else:\n",
        "                freqs[i][k+1][base] = freqs[i][k+1][base]/freqs[i][k][base[:-1]]# переходная вероятность: freqs[AC]/freqs[A]\n",
        "        # Считаем правдоподобие\n",
        "        for j in range(k, len(sequence[i])-k):\n",
        "            L += log(freqs[i][k+1][sequence[i][j:j+k+1]])\n",
        "        AIC += calc_AIC(L, s)\n",
        "        BIC += calc_BIC(L, s)\n",
        "    return {'Model' : f'Markov_{k}',\n",
        "            'AIC' : round(AIC),\n",
        "            'BIC' : round(BIC)}"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0AUP4DyMH-Zo"
      },
      "source": [
        ">#### 4. Для каждой модели оценить AIC и BIC."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AYxRp5QC9D1I"
      },
      "source": [
        "final = [Bernoulli(sequences),]"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uvcuvuuK3ir7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "dde481bd-9199-4e5a-e610-ff0510c7e960"
      },
      "source": [
        "for k in tqdm(range(1,5)):\n",
        "    final.append(Markov(sequences,k=k))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "  0%|          | 0/4 [00:00<?, ?it/s]\u001b[A\n",
            " 25%|██▌       | 1/4 [00:16<00:49, 16.44s/it]\u001b[A\n",
            " 50%|█████     | 2/4 [00:34<00:33, 16.99s/it]\u001b[A\n",
            " 75%|███████▌  | 3/4 [00:58<00:19, 19.00s/it]\u001b[A\n",
            "100%|██████████| 4/4 [02:58<00:00, 44.55s/it]\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "kmBc5Xs1C2RS",
        "outputId": "06d7aad9-e568-42d4-ee5d-06f72cca849e"
      },
      "source": [
        "df = pd.DataFrame(final, index = [i for i in range(1, 6)])\n",
        "df"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Model</th>\n",
              "      <th>AIC</th>\n",
              "      <th>BIC</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Bernoulli</td>\n",
              "      <td>11064948</td>\n",
              "      <td>11079671</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Markov_1</td>\n",
              "      <td>10958569</td>\n",
              "      <td>11032185</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Markov_2</td>\n",
              "      <td>10892374</td>\n",
              "      <td>11201563</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Markov_3</td>\n",
              "      <td>11041494</td>\n",
              "      <td>12292972</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Markov_4</td>\n",
              "      <td>11864421</td>\n",
              "      <td>16885055</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "       Model       AIC       BIC\n",
              "1  Bernoulli  11064948  11079671\n",
              "2   Markov_1  10958569  11032185\n",
              "3   Markov_2  10892374  11201563\n",
              "4   Markov_3  11041494  12292972\n",
              "5   Markov_4  11864421  16885055"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 124
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mDiKnIqJIZju"
      },
      "source": [
        ">#### 5. Сделать вывод – какая модель лучше описывает последовательности."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qd1NN7rBFmM3",
        "outputId": "de5d4efe-2300-49dd-d5ef-e303899d726e"
      },
      "source": [
        "min_AIC = df['AIC'].idxmin()\n",
        "min_BIC = df['BIC'].idxmin()\n",
        "print(f'Best model by AIC | {df.Model[min_AIC]}\\n\\nBest model by BIC | {df.Model[min_BIC]}')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Best model by AIC | Markov_2\n",
            "\n",
            "Best model by BIC | Markov_1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SDU3i8oEJ768"
      },
      "source": [
        ">Если судить по критерию AIC лучше всего фрагменты генома *E.coli* описывает марковская модель 2 порядка.<br/>\n",
        "Если же смотреть по критерию BIC, то лучшая - марковская модель 1 порядка.\n",
        "Исходя из этого, наверное, можно судить о \"триплетности\" генома"
      ]
    }
  ]
}